soroban-sdk/src/env.rs

```diff
@@ -576,6 +578,91 @@ impl Env {
         env
     }
 
+    /// Enables detailed metering of contract invocations.
+    ///
+    /// The top-level contract invocations and lifecycle operations (such as
+    /// `register` or `env.deployer()` operations) will be metered and the
+    /// budget will reset in-between them. Metering will not be reset while
+    /// inside the call (e.g. if a contract calls or creates another contract,
+    /// that won't reset metering).
+    ///
+    /// The metered resources for the last invocation can be retrieved with
+    /// `get_last_invocation_resources()` and the estimated fee corresponding
+    /// to these resources can be estimated with
+    /// `estimate_last_invocation_fee()`.
+    ///
+    /// While the resource metering may be useful for contract optimization,
+    /// keep in mind that resource and fee estimation may be imprecise. Use
+    /// simulation with RPC in order to get the exact resources for submitting
+    /// the transactions to the network.
+    pub fn enable_invocation_metering(&self) {
+        self.env_impl.enable_invocation_metering();
+    }
+
+    /// Returns the resources metered during the last top level contract
+    /// invocation.    
+    ///
+    /// In order to get non-`None` results, `enable_invocation_metering` has to
+    /// be called and at least one invocation has to happen after that.
+    ///
+    /// Take the return value with a grain of salt. The returned resources mostly
+    /// correspond only to the operations that have happened during the host
+    /// invocation, i.e. this won't try to simulate the work that happens in
+    /// production scenarios (e.g. certain XDR rountrips). This also doesn't try
+    /// to model resources related to the transaction size.
+    ///
+    /// The returned value is as useful as the preceding setup, e.g. if a test
+    /// contract is used instead of a Wasm contract, all the costs related to
+    /// VM instantiation and execution, as well as Wasm reads/rent bumps will be
+    /// missed.    
+    pub fn get_last_invocation_resources(&self) -> Option<InvocationResources> {
```

Comment:
I'd like to move the functions for getting the resources used into some other type which is available on Env, because otherwise Env becomes a bit of a function soup over time.

We currently have `env.budget()` that returns a testutils `Budget` (defined in this repo, not the same as the env's Budget). Should these functions replace that? Or can we reimagine both together in a single location? Budget getting the raw usage, and this doing something else? From best I can tell looks like this replaces the existing Budget object.

Reply:
> From best I can tell looks like this replaces the existing Budget object.

Not quite, `Budget` provides the detailed breakdown of the CPU costs, while this provides a high level overview of all the resources used. I think both have a place in development process.

That said, I agree that we could try converging this functionality at the SDK level, especially given that with this change everything will be describing the last invocation only.

Reply:
That seems fine, though I'd like to emphasize that these are estimates so it's hopefully not as tempting to use this instead of simulation. So I ended up with `cost_estimate()`.

I've rebased everything from scratch as I was debugging some silly issues, PTAL.

Reply:
> So I ended up with `cost_estimate()`.

👍🏻 I think that's wise to call out their estimates.